src/util/sighash.rs

```diff
@@ -0,0 +1,836 @@
+            _ => Err(Error::PrevoutKind),
+        }
+    }
+
+    fn get(&self, input_index: usize) -> Result<&TxOut, Error> {
+        match self {
+            Prevouts::One(index, prevout) => {
+                if input_index == *index {
+                    Ok(prevout)
+                } else {
+                    Err(Error::PrevoutIndex)
+                }
+            }
+            Prevouts::All(prevouts) => prevouts.get(input_index).ok_or(Error::PrevoutIndex),
+        }
+    }
+}
+
+impl<'s> ScriptPath<'s> {
+    /// Create a new ScriptPath structure
+    pub fn new(script: &'s Script, code_separator_pos: u32, leaf_version: u8) -> Self {
+        ScriptPath {
+            script,
+            code_separator_pos,
+            leaf_version,
+        }
+    }
+    /// Create a new ScriptPath structure using default values for `code_separator_pos` and `leaf_version`
+    pub fn with_defaults(script: &'s Script) -> Self {
+        Self::new(script, 0xFFFFFFFFu32, LEAF_VERSION_TAPSCRIPT)
+    }
+}
+
+impl From<LegacySigHashType> for SigHashType {
+    fn from(s: LegacySigHashType) -> Self {
+        match s {
+            LegacySigHashType::All => SigHashType::All,
+            LegacySigHashType::None => SigHashType::None,
+            LegacySigHashType::Single => SigHashType::Single,
+            LegacySigHashType::AllPlusAnyoneCanPay => SigHashType::AllPlusAnyoneCanPay,
+            LegacySigHashType::NonePlusAnyoneCanPay => SigHashType::NonePlusAnyoneCanPay,
+            LegacySigHashType::SinglePlusAnyoneCanPay => SigHashType::SinglePlusAnyoneCanPay,
+        }
+    }
+}
+
+impl SigHashType {
+    /// Break the sighash flag into the "real" sighash flag and the ANYONECANPAY boolean
+    pub(crate) fn split_anyonecanpay_flag(self) -> (SigHashType, bool) {
+        match self {
+            SigHashType::Default => (SigHashType::Default, false),
+            SigHashType::All => (SigHashType::All, false),
+            SigHashType::None => (SigHashType::None, false),
+            SigHashType::Single => (SigHashType::Single, false),
+            SigHashType::AllPlusAnyoneCanPay => (SigHashType::All, true),
+            SigHashType::NonePlusAnyoneCanPay => (SigHashType::None, true),
+            SigHashType::SinglePlusAnyoneCanPay => (SigHashType::Single, true),
+            SigHashType::Reserved => (SigHashType::Reserved, false),
+        }
+    }
+}
+
+impl<R: Deref<Target = Transaction>> SigHashCache<R> {
+    /// Compute the sighash components from an unsigned transaction and auxiliary
+    /// in a lazy manner when required.
+    /// For the generated sighashes to be valid, no fields in the transaction may change except for
+    /// script_sig and witnesses.
+    pub fn new(tx: R) -> Self {
+        SigHashCache {
+            tx,
+            common_cache: None,
+            taproot_cache: None,
+            segwit_cache: None,
+        }
+    }
+
+    fn check_index(&self, index: usize) -> Result<(), Error> {
```

Comment:
Implementing this helped me find a validation bug in the case of SIGHASH_SINGLE when there is no corresponding output, according to the bip341 spec it's a validation error. 
Instead, it was allowed (and bound checked) in segwit by looking at the code.

Note that I left a check in the legacy case because I don't want to change the signature of the method in [Transaction::encode_signing_data_to] but I want panic-free code here

done in 4c1aaa6